pallets/parachain-system/src/lib.rs

```diff
@@ -296,6 +299,34 @@ pub mod pallet {
 				weight += T::DbWeight::get().writes(1);
 			}
 
+			let para_head = ParaHead::<T>::take();
+			weight += T::DbWeight::get().reads_writes(1, 1);
+
+			// Update unincluded segment related storage values.
+			if let Some(para_head) = para_head {
```

Comment:
`para_head` should not be allowed to be `None` once the parachain has been configured to support asynchronous backing. Otherwise, collators could simply choose to omit the `para_head` from the proof and break the logic.

We should have a 'switch' for parachain governance to enable the unincluded-segment functionality; after that point, it should be invalid for the relay-chain storage proof _not_ to contain the para head.

Reply:
It should also be invalid for a parachain block to execute `on_initialize` when the unincluded segment is full and has not had anything drained.

Reply:
We can introduce a config value for `MaxUnincludedLen`. If it's non-zero, then the switch is on

Reply:
I think the best thing is a storage item which can be set by governance. There are situations where you might want `MaxUnincludedLen` to be 0 but use the relay-chain state proof. Maybe an `Option<BlockNumber>` is best - it starts as `None`, but there should be no way to set it back to `None` once it's `Some`.

Reply:
i.e. @burdges suggested that we should have pluggable logic for the relay chain to set max-unincluded-len dynamically during on-initialize based on the relay-chain state, like how many cores are available, how far in the future, etc.

Reply:
[cdf3ba9](https://github.com/paritytech/cumulus/pull/2438/commits/cdf3ba9f0a6b44e43c72406df0eaa063f3fb4521)